test/test_basic_logic.py

```diff
@@ -850,6 +850,85 @@ def test_changing_max_frame_size(self, frame_factory, frame_size):
         with pytest.raises(h2.exceptions.ProtocolError):
             c.receive_data(data.serialize())
 
+    def test_cookies_are_joined_on_push(self, frame_factory):
+        """
+        RFC 7540 Section 8.1.2.5 requires that we join multiple Cookie headers
+        in a header block together when they're received on a push.
+        """
+        # This is a moderately varied set of cookie headers: some combined,
+        # some split.
+        cookie_headers = [
+            ('cookie',
+                'username=John Doe; expires=Thu, 18 Dec 2013 12:00:00 UTC'),
+            ('cookie', 'path=1'),
+            ('cookie', 'test1=val1; test2=val2')
+        ]
+        expected = (
+            'username=John Doe; expires=Thu, 18 Dec 2013 12:00:00 UTC; '
+            'path=1; test1=val1; test2=val2'
+        )
+
+        c = h2.connection.H2Connection()
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers, end_stream=True)
+
+        f = frame_factory.build_push_promise_frame(
+            stream_id=1,
+            promised_stream_id=2,
+            headers=self.example_request_headers + cookie_headers
+        )
+        events = c.receive_data(f.serialize())
+
+        assert len(events) == 1
+        e = events[0]
+
+        cookie_field_count = sum(1 for n, _ in e.headers if n == 'cookie')
+        assert cookie_field_count == 1
+
+        for n, v in e.headers:
+            if n == 'cookie':
+                assert v == expected
+                break
+        else:
+            pytest.mark.fail("Unable to check cookie header")
```

Comment:
When would we ever come down the `else:` branch of this loop?

If we pass the assert above, we know that there’s exactly one header whose name is `cookie`. So unless `e.headers` changes in the second iteration, we’ll always run into the `break` statement.

How about something like this?

```python
cookie_fields = [(n, v) for (n, v) in e.headers if n == 'cookie']
assert len(cookie_fields) == 1

_, v = cookie_fields[0]
assert v == expected
```